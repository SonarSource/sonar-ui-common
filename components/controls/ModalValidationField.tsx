/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import AlertErrorIcon from '../icons/AlertErrorIcon';
import AlertSuccessIcon from '../icons/AlertSuccessIcon';

interface Props {
  children: (props: { className?: string }) => React.ReactNode;
  description?: string;
  dirty: boolean;
  error: string | undefined;
  label?: React.ReactNode;
  touched: boolean | undefined;
}

export default function ModalValidationField(props: Props) {
  const { description, dirty, error } = props;

  const isValid = dirty && props.touched && error === undefined;
  const showError = dirty && props.touched && error !== undefined;
  return (
    <div className="modal-validation-field">
      {props.label}
      {props.children({ className: classNames({ 'is-invalid': showError, 'is-valid': isValid }) })}
      {showError && <AlertErrorIcon className="little-spacer-top" />}
      {isValid && <AlertSuccessIcon className="little-spacer-top" />}
      {showError && <p className="text-danger">{error}</p>}
      {description && <div className="modal-field-description">{description}</div>}
    </div>
  );
}
