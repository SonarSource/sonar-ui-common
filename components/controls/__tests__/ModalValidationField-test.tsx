/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import ModalValidationField from '../ModalValidationField';

it('should display the field without any error/validation', () => {
  expect(getWrapper({ description: 'Describe Foo.', touched: false })).toMatchSnapshot();
  expect(getWrapper({ dirty: false })).toMatchSnapshot();
});

it('should display the field as valid', () => {
  expect(getWrapper({ error: undefined })).toMatchSnapshot();
});

it('should display the field with an error', () => {
  expect(getWrapper()).toMatchSnapshot();
});

function getWrapper(props = {}) {
  return shallow(
    <ModalValidationField
      dirty={true}
      error="Is required"
      label={<label>Foo</label>}
      touched={true}
      {...props}>
      {({ className }) => <input className={className} type="text" />}
    </ModalValidationField>
  );
}
