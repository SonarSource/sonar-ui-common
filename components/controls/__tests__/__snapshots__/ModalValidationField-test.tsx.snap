// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display the field as valid 1`] = `
<div
  className="modal-validation-field"
>
  <label>
    Foo
  </label>
  <input
    className="is-valid"
    type="text"
  />
  <AlertSuccessIcon
    className="little-spacer-top"
  />
</div>
`;

exports[`should display the field with an error 1`] = `
<div
  className="modal-validation-field"
>
  <label>
    Foo
  </label>
  <input
    className="is-invalid"
    type="text"
  />
  <AlertErrorIcon
    className="little-spacer-top"
  />
  <p
    className="text-danger"
  >
    Is required
  </p>
</div>
`;

exports[`should display the field without any error/validation 1`] = `
<div
  className="modal-validation-field"
>
  <label>
    Foo
  </label>
  <input
    className=""
    type="text"
  />
  <div
    className="modal-field-description"
  >
    Describe Foo.
  </div>
</div>
`;

exports[`should display the field without any error/validation 2`] = `
<div
  className="modal-validation-field"
>
  <label>
    Foo
  </label>
  <input
    className=""
    type="text"
  />
</div>
`;
